orchestrator/relayer/src/fee_manager.rs

```diff
@@ -0,0 +1,219 @@
+    pub async fn can_send_batch(
+        &mut self,
+        estimated_cost: &GasCost,
+        batch_fee: &Erc20Token,
+        contract_address: &EthAddress,
+    ) -> bool {
+        if self.mode == RelayerMode::AlwaysRelay {
+            return true;
+        }
+
+        match self.next_batch_send_time.get(contract_address) {
+            Some(time) => {
+                if *time < Instant::now() {
+                    return true;
+                }
+            }
+            None => self.update_next_batch_send_time(*contract_address),
+        }
+
+        let token_price = match self.get_token_price(&batch_fee.token_contract_address).await {
+            Ok(token_price) => token_price,
+            Err(_) => return false,
+        };
+
+        let estimated_fee = estimated_cost.get_total();
+        let batch_value = batch_fee.amount.clone() * token_price;
+
+        info!("estimate cost is {}, batch value is {}", estimated_fee, batch_value);
+        batch_value >= estimated_fee
+    }
+
+    pub fn update_next_batch_send_time(
+        &mut self,
+        contract_address: EthAddress,
+    ) {
+        if self.mode == RelayerMode::AlwaysRelay {
+            return;
+        }
+
+        let timeout_duration = std::env::var("GRAVITY_BATCH_SENDING_SECS")
+            .map(|value| Duration::from_secs(value.parse().unwrap()))
+            .unwrap_or_else(|_| Duration::from_secs(3600));
+
+        self.next_batch_send_time.insert(contract_address, Instant::now() + timeout_duration);
+    }
+
+    async fn get_token_price(&mut self, contract_address: &EthAddress) -> Result<U256, ()> {
+        match self.mode {
+            RelayerMode::Api => {
+                if self.token_api_url_map.contains_key(&*format_eth_address(*contract_address)) {
+                    let api_url = self.token_api_url_map
+                        .get(&*format_eth_address(*contract_address));
+
+                    if api_url.is_none() {
+                        log::error!("Api url expected to be string format");
+                        return Err(());
+                    }
+                    let request_url = api_url.unwrap();
+
+                    // Return because gas price is quoted in ETH
+                    if request_url == "ETH" {
+                        return Ok(U256::from(1))
+                    }
+
+                    let response = reqwest::get(request_url).await.expect("Cannot parse response from oracle");
+                    let result : ApiResponse = response.json().await.expect("Cannot parse result from oracle");
+                    // TODO to be updated with new oracle API
+                    let token_price = result.result.aggregatedPrice * (i32::pow(10, result.result.tokenPairDecimal) as f64);
+                    let token_price_u64=  U256::from_dec_str(&token_price.to_string()).expect("Cannot convert token price");
```

Comment:
we can use `map_err(|e| ...)?` instead of `expect`, `expect` will crush the program.

Reply:
still learning about rust. do you have an example how to use the map_err?

Reply:
for example:
```
let response = reqwest::get(request_url).await.map_err(|e| {
    log::error("request error: {:?}", e);
    ()
}?;
```
the function you returned `Result<U256, ()>`, where `U256` is what we want, and `()` is the error type, you can use another `Error` type (for example `anyhow::Error`, or some other `Error` in the project)  instead of `()`, because there is no other error information in `()`, when call `get_token_price` and it returns `Error`, all error is `()`, it's not good.